src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class DecisionDagBuilder
+    {
+        private void MakeTestsAndBindingsForLengthAndListPatterns(BoundDagTemp input, BoundRecursivePattern recursive, ArrayBuilder<BoundPatternBinding> bindings, ArrayBuilder<Tests> tests)
+        {
+            if (recursive.HasErrors)
+                return;
+
+            Debug.Assert(recursive.LengthProperty is not null || input.Type.IsSZArray());
+            var syntax = recursive.Syntax;
+            var lengthProperty = recursive.LengthProperty ?? (PropertySymbol)_compilation.GetSpecialTypeMember(SpecialMember.System_Array__Length);
+            var lengthEvaluation = new BoundDagPropertyEvaluation(syntax, lengthProperty, input);
+            tests.Add(new Tests.One(lengthEvaluation));
+            var lengthTemp = new BoundDagTemp(syntax, _compilation.GetSpecialType(SpecialType.System_Int32), lengthEvaluation);
+
+            if (recursive.LengthPattern is not null)
+            {
+                tests.Add(MakeTestsAndBindings(lengthTemp, recursive.LengthPattern, bindings));
+            }
+
+            if (recursive.ListPatternClause is { Subpatterns: var subpatterns } clause)
+            {
+                Debug.Assert(!subpatterns.IsDefaultOrEmpty);
+
+                tests.Add(new Tests.One(clause.HasSlice
+                    ? new BoundDagRelationalTest(syntax, BinaryOperatorKind.IntGreaterThanOrEqual, ConstantValue.Create(subpatterns.Length - 1), lengthTemp)
+                    : new BoundDagValueTest(syntax, ConstantValue.Create(subpatterns.Length), lengthTemp)));
+
+                for (int index = 0; index < subpatterns.Length; index++)
+                {
+                    var subpattern = subpatterns[index];
+                    if (subpattern is BoundSlicePattern slice)
+                    {
+                        if (slice.Pattern is not null)
+                        {
+                            var sliceEvaluation = new BoundDagSliceEvaluation(syntax, slice.SliceType, lengthTemp, startIndex: index, endIndex: index - (subpatterns.Length - 1), slice.IndexerInfo, input);
+                            tests.Add(new Tests.One(sliceEvaluation));
+                            var sliceTemp = new BoundDagTemp(syntax, slice.SliceType, sliceEvaluation);
+                            tests.Add(MakeTestsAndBindings(sliceTemp, slice.Pattern, bindings));
+                        }
+
+                        for (int i = subpatterns.Length - 1, j = -1; i > index; i--, j--)
+                        {
+                            addIndexerTests(index: j, subpatterns[i]);
```

Comment:
We're starting with index `-1` and proceeding downards. I think we should stick with syntactic order (proceeding upwards) instead.

Reply:
This would generate a slightly better DAG (all trailing matches would be merged together). I think that's ok because the order of evaluation is undefined here and may be optimized one way or another. 
I can revert if you feel that's *out of place*.

Reply:
> This would generate a slightly better DAG (all trailing matches would be merged together). 

I didn't follow that part.
It's true that the order of evaluation is intentionally left unspecified, but I don't see a reason for the implementation to choose this order.
I'm suggesting that the order of evaluation for `{ 1, 2, .., -2, -1 }` be: `[0]`, `[1], `Slice(2, -2)`, `[^2]`, `[^1]`. But right now it is ... `[^1]`, `[^2]` (last two index accesses are swapped) which seems odd and unnecessary.